src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php

```diff
@@ -49,82 +51,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
     public function handle(AddCartRuleToOrderCommand $command): void
     {
         $order = $this->getOrderObject($command->getOrderId());
+        $orderInvoice = $this->createSingleOrderInvoice($order, $command);
```

Comment:
It would be clearer if you managed the `single order invoice` case here, if `createSingleOrderInvoice` return null it means you have to apply the discount to all its invoices. So you could do:
```php
$orderInvoice = $this->createSingleOrderInvoice($order, $command);
if (null !== $orderInvoice) {
    $orderInvoices = [$orderInvoice];
} else {
    $orderInvoices = $order-> getInvoicesCollection()->getAll();
}
```

This way you always have an array of OrderInvoice objects, and you can simplify most of your methods, you don't need to transfer the Order and a potentially null OrderInvoice. The method `buildCartRulesByType` (and its subsequent methods) can be simplified into `buildCartRulesByType($orderInvoices, $discountValue, $cartRuleType)`